blockchain/client.go

```diff
@@ -129,80 +135,74 @@ func (c *Client) ListenEvents(
 
 	// Sequence historical and live headers.
 	headersC := make(chan types.Header, 2)
-	g.Go(func() error {
-		defer close(headersC)
-
-		for header := range histHeadersC {
-			select {
-			case <-ctx.Done():
-				return ctx.Err()
-			case headersC <- header:
-			}
-		}
+	defer close(headersC)
 
-		for header := range liveHeadersC {
-			select {
-			case <-ctx.Done():
-				return ctx.Err()
-			case headersC <- header:
-			}
+	g.Go(func() error {
+		if err = forwardHeaders(ctx, histHeadersC, headersC); err != nil {
+			return err
 		}
 
-		return nil
+		return forwardHeaders(ctx, histHeadersC, headersC)
 	})
 
 	// Retrieve events skipping blocks before 'begin'.
 	eventsC := make(chan blockEvents, 2)
-	g.Go(func() error {
-		defer close(eventsC)
-
-		for header := range headersC {
-			if header.Number < begin {
-				continue
-			}
-
-			hash, err := c.RPC.Chain.GetBlockHash(uint64(header.Number))
-			if err != nil {
-				return err
-			}
-
-			events, err := retriever.GetEvents(hash)
-			if err != nil {
-				return err
-			}
+	defer close(eventsC)
 
+	g.Go(func() error {
+		for {
 			select {
 			case <-ctx.Done():
 				return ctx.Err()
-			case eventsC <- blockEvents{
-				Events: events,
-				Hash:   hash,
-				Number: header.Number,
-			}:
+			case header := <-headersC:
```

Comment:
Should we check it with `item, ok := <-C` here as well as in other places?

Reply:
that is why i moved closure outside of those goroutines.